tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc

```diff
@@ -120,6 +120,171 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
   return true;
 }
 
+namespace {
+bool Is1dSliceWithoutStrides(const HloInstruction* instr) {
+  return instr->opcode() == HloOpcode::kSlice &&
+         1 == instr->slice_starts().size() &&
+         1 == instr->slice_limits().size() &&
+         1 == instr->slice_strides().size() &&
+         1 == instr->slice_strides().at(0);
+}
+
+bool IsSliceInputFusion(const HloInstruction& unnested_hlo) {
+  if (!unnested_hlo.IsInputFusion()) {
+    return false;
+  }
+  const HloInstruction* root = unnested_hlo.fused_expression_root();
+  if (root->opcode() != HloOpcode::kTuple) {
+    return false;
+  }
+  return absl::c_all_of(root->operands(), [](const HloInstruction* instr) {
+    return Is1dSliceWithoutStrides(instr);
+  });
+}
+
+// Returns whether the concat is used in an elementwise manner.
+// A concat followed by slices is considered (effectively) elementwise if the
+// slices combinedly is a reverse function of the concat.
+// prev_concat_opnd_idx: previously seen concat and its operand index.
+//     prev_concat_opnd_idx.first is nullptr if no previously seen concat.
+bool ConcatIsEffectivelyElementwise(
```

Comment:
Style nit: return value is preferred over output parameters.

Can we put the inout values in a struct? That would also give a better chance to explain their semantics.

E.g. something like

```
struct ConcatUsageInfo {
  const HloInstruction* concat;
  int64 op_idx;
  const HloInstruction* slice;
};

absl::optional<ConcatUsageInfo> ConcatIsEffectivelyElementwise(..., const ConcatUsageInfo& info);
```

?

Reply:
Revised accordingly. Thanks for the suggestion.